c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -153,21 +158,23 @@ static std::string format_size(uint64_t size) {
 
 } // namespace
 
-struct THCCachingAllocator
-{
-  // device statistics
-  std::vector<DeviceStats> device_stats;
+class THCCachingAllocator {
+
+ private:
 
   // lock around all operations
-  std::recursive_mutex mutex;
+  mutable std::mutex mutex;
```

Comment:
Gotcha – yep didn't realize that the callback functionality could/would re-enter caching allocator. thanks for catching!

Reply:
Actually, now that I think about it, I'm not sure why changing from recursive_mutex to mutex would result in a /stochastic/ failure instead of all-or-nothing – a deadlock on a single thread trying to re-acquire a non-recurrent lock should be deterministic.

Regardless, this is a minor point and I think recursive_mutex is fine here.